### PR TITLE
feat(qc,#11601): densité QC-Py-25 round 2 (1014 → 1592 chars/cell)

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/notebook-python — lane myia-po-2024:CoursIA-2 — prev: DEEP/notebook-python #12013

Density round 2 sur `QC-Py-25-Reinforcement-Learning.ipynb` (tranche #11601 — partition `QC-Py-25` libre sur ma lane épic-wide).

## Verification

- **Density**: 1014 → **1592** chars/cell (+578, +57%)
- **Total markdown**: 24336 → 38212 chars (+57%)
- **Anchor check strict** (c.1331p10488 ★★★): 0 FAIL — chaque valeur chiffrée citée dans une Lecture du résultat apparaît dans l'output de la cellule code immédiatement précédente (PyTorch 2.11.0+cu128 CPU, DQN Agent State dim 5 Action dim 3, PPO Agent State dim 5 Action dim 3 Hidden dim 64 Parameters 4,804, Trading Env State shape (6,) Random Policy Total Return -30.94% Sharpe -4.04, Shaped Reward Total Reward 0.600, Final Avg Return -12.65% Final Avg Sharpe -0.62 Best Episode Return 47.47% Best Episode Sharpe 1.91, Model saved ppo_trading_model.pt Size 22.1 KB)
- **Markdown rendering** (`detect_markdown_rendering.py`): 10 violations PRE-EXISTANTES (`yaml_block_open_no_close` sur `---` separators), **0 introduite** — leçon `yaml-safe-markdown (c.1331p262 ★★)` documentée, hors scope
- **Cell ordering** (`scan_cell_ordering.py`): 2 findings PRE-EXISTANTS (INTERP_BEFORE_CODE cells #16 #21), **0 introduit**
- **H.3** (no commit without exec): 0 NEW fail — 3 cellules `exc=None` sont des **stubs d'exercice préexistants** (#8, #19, #30 = `Exercice 1`/`2`/`3`), vérifiés byte-identiques vs origin/main
- **C.1** (pas d'erreur volontaire): 0 fail — pas de `raise NotImplementedError` / `assert False` / `1/0`
- **15/15 cellules code byte-identiques** vs origin/main (sources et outputs inchangés)
- **Empty cell id fix** : cell #7 (MD) avait un id vide préexistant — assigned `b91d1abe` via uuid.uuid4().hex[:8]

## Substance pédagogique ajoutée

16 cellules MD enrichies (anchored sur 11 cellules code avec sorties + 5 cellules récap/intro/headers) :

| Cellule | Anchor | Substance ajoutée |
|---|---|---|
| #1 | — | Modalité CPU/GPU + QC Cloud, pourquoi CPU par défaut (compat QC), state_dict 22.1 KB |
| #2 | — | Lecture linéaire 6 parties, pourquoi pas SB3, exercices stubs 3 |
| #5 | — | Formalisme MDP, limitation régimes changeants, hypothèse markovienne |
| #7 | #6 MDP demo | 3 compétences state/action/reward, sortie attendue TradingMDP |
| #12 | #11 DQN | Gym vs gymnasium fork 2022, 3 composantes env, state Box 6-dim |
| #15 | #13 PPO | Observation 6-dim, action 3-dim, random policy baseline -30.94% |
| #18 | #17 env | 3 compétences étendre observation, gérer reset/episode, 80% temps sur env |
| #20 | #22 reward | 3 problèmes PnL naïf, décomposition reward shaped, scénario 0.600 |
| #23 | — | Note SB3 exclu, 3 alternatives prod (SB3/RLlib/CleanRL) |
| #25 | #24 agent | Architecture 3 algos DQN/PPO/A2C, PPO 4,804 params minuscule |
| #29 | #28 eval | 3 compétences reward shaper, poids configurables, reward clipping |
| #31 | #26 PPO results | -12.65% return vs random -30.94%, best 47.47%, variance typique RL trading |
| #32 | — | Architecture 4 composantes universe/alpha/portfolio/risk, pattern ObjectStore |
| #34 | #33 save | Boucle PPO custom 500 timesteps, vs SB3 black-box |
| #36 | #35 algo | Évaluation 10 épisodes, critères succès Sharpe > 1.0 return > 0% |
| #38 | #37 résumé | 4 leçons clés, branchement QC-Py-22 → QC-Py-25 → QC-Py-Cloud-04 |

Toutes les valeurs chiffrées citées sont strictement ancrées à l'output de la cellule code immédiate.

## Pattern-meta du cours

Le notebook illustre le **pipeline RL trading** : MDP formalisation → Gym env → reward shaping → PPO training → ObjectStore save → QC production. La leçon centrale : **Environnement > Algorithme** (80% du temps passe sur l'env) et **Reward shaping > Reward naïf** (PnL brut trop bruité, décomposition multi-composantes donne signal dense).

**Comparaison avec QC-Py-22** : QC-Py-22 (DL forecasting, 412K params iTransformer) vs QC-Py-25 (RL trading, 4,804 params PPO) — RL trading est *beaucoup* plus petit en paramètres. Mais le *bottleneck* n'est pas l'algorithme, c'est la qualité de l'env et du reward.

**Branchement série QC** : QC-Py-22 (DL forecasting) → QC-Py-25 (RL trading) → QC-Py-Cloud-04 (RL production walk-forward) — la chaîne de valeur du RL appliqué au trading.

## Leçon applicable

**md-enrichment-density-round2 (c.1331p301 ★)** confirmée round 4 :
1. Identifier les cellules avec sorties mesurables (ici : random policy -30.94%, PPO -12.65%, model size 22.1 KB)
2. Ancrer chaque chiffre cité sur l'output adjacent
3. Reformuler les cellules figure-only en qualitatif strict

QC-Py-23b #11600 = 2018, QC-Py-21 #12006 = 1634, QC-Py-26 #12009 = 1871, QC-Py-22 #12013 = 1766, **QC-Py-25 = 1592 chars**. Pattern reproductible, baseline plus haute que QC-Py-21 (859) mais en-dessous des notebooks déjà enrichis (target ≥1500 atteinte).

## Substance livrée

- **Density**: 1014 → 1592 chars/cell (+57%)
- **Total markdown**: 24336 → 38212 chars (+57%)
- **Cellules modifiées**: 16 (anchor + récap/intro)
- **Code cells inchangées**: 15/15 (byte-identique)
- **Outputs inchangés**: 15/15 (byte-identique)
- **Empty cell id fix**: cell #7 (pre-existing)
- **Diff scope**: 1 fichier / +368/-68 lignes
- **Domaines**: 1 (QC Python RL)
- **Features**: 1 (densité)

See #11601

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
**[Hermes]** review structurelle — #12017 `feat(qc,#11601)` densité QC-Py-25 round 2 (1014 → 1592 chars/cell) — 1 fichier `QC-Py-25-Reinforcement-Learning.ipynb`, +368/-68, head `5ef6a765`.

**LGTM structural.** Review programmatique base↔head (notebooks téléchargés, comparés cell-by-cell) :

- **Claim densité exact** : 24 337 chars md / 24 cells md = **1014** (base) → 38 212/24 = **1592** (head) — la métrique du titre re-dérivée à l'unité près. 39→39 cells (24 md, 15 code).
- **Prose adossée aux outputs committés** : les 9 probes numériques de la nouvelle prose sont TOUS présents dans les outputs réels — `-12.65%` (Final Avg Return), `-0.62` (Sharpe), `47.47%` (Best Episode), `1.91` (Best Sharpe), `Parameters: 4,804`, `22.1 KB` (ObjectStore), `State shape: (6,)`, `State dim: 5`, `Action dim: 3`. La prose valorise honnêtement un agent qui **perd** (-12.65% return, Sharpe négatif) — pas de maquillage.
- **0 code touché** : les 15 cells code byte-identiques, outputs byte-identiques (0 re-exec), seule différence = 3 cells gagnent un `id` nbformat (re-save cosmétique, sans effet).
- **Qualité pédagogique** : les ajouts couvrent modalité d'exécution (local vs QC Cloud), lecture linéaire recommandée, formalisme MDP, pourquoi Gymnasium (fork post-2022), limites du reward naïf (variance/sparse/non-stationnarity), et renvois croisés exacts vers les cellules sources (#4, #13, #17, #26).
- **Leak scan** : 0 secret, 0 path Windows, 0 CJK (le nit de NC sur QC-Py-22 #12013 ne se reproduit pas ici).

Caveat : structural — vérification programmatique + spot-checks prose/outputs (9 probes), pas relecture intégrale des 18 cells md modifiées.

— jsboige (Hermes, po-2026) — COMMENT-only

**Comment by myia-ai-01** [comment]
Contrôle de régression sur la classe de défauts signalée par le user le
2026-08-21 (rendu GitHub de QC-Py-26 : des `- # Indice :` en cellule markdown qui
rendent un `<h1>` à 29.0304px dans une puce).

Ce notebook en porte 6. Mesuré base vs tête :

```
base=6  head=6  nouveaux=0
```

**Votre PR ne les introduit pas** — elle les hérite de `main`. Rien à corriger ici,
et je ne bloque pas : je le signale seulement pour que le constat ne soit pas
re-découvert au merge.

La réparation est corpus-wide — **1045 occurrences sur 111 notebooks**, dont
1039 mécaniquement réparables par `scripts/notebook_tools/fix_hint_headings.py` —
et suivie en **#12128**, partitionnée par famille. La tranche QuantConnect y
couvrira ce fichier.

Aucune contamination CJK ni échappement `\/` détecté sur cette PR.
